File: adapters/companion.py
# /// script
# requires-python = ">=3.11"
# dependencies = [
#     "homeostat",
#     "paho-mqtt>=2,<3",
# ]
#
# [tool.uv.sources]
# homeostat = { path = "../sdk/python", editable = true }
# ///
"""Companion adapter: the house's own Android app, over the MQTT broker
(docs/design.md, Notifications; issue #51).

The membrane that makes a family phone a described part of the house. The
phone never touches the bus — it is a device, and devices speak a dialect
to an adapter (docs/design.md, Bus). Unlike every other dialect this one
is ours, so the app already speaks homeostat vocabulary and there is
nothing to rename; what this adapter provides is the ownership, the
grants and the plan visibility that a phone publishing for itself would
not have.

Each phone owns a topic subtree under the endpoint's base topic
(`companion` unless the endpoint path names another) and two entities:

    {base}/{phone}/available          <- retained birth `true`, LWT `false`
    {base}/{phone}/person/presence    <- bare JSON bool
    {base}/{phone}/person/position    <- {"lat","lon","accuracy","battery","fixed_at"}
    {base}/{phone}/notifier/ack       <- bare JSON epoch seconds
    {base}/{phone}/notifier/message   -> {"text","actor","sent_at"}, QoS 1
    {base}/{phone}/notifier/alert     -> the same, the urgent channel

The entity file's `id` is the two topic segments that name the subtree and
the face: "alice/person" binds the `person` entity (room `person`, the
reserved pseudo-room), "alice/notifier" the `notifier` entity for the same
phone. One phone is therefore two files, which is what the two
capabilities are; they share the first segment and that is how `available`
reaches both.

Published state:

- `presence` (bool) on the person, from the platform's geofence
  transitions. The app registers ONE geofence, `home`; continuous position
  is opt-in per phone and off by default (#51), so a phone that never
  publishes `position` is the normal case, not a degraded one.
- `lat`, `lon` and — when the fix carries them — `accuracy`, `battery`,
  `fixed_at`: scalar aspects, never one composite fix, so the recorder
  gives trails for free.
- `delivered` (epoch seconds) on the notifier when the BROKER acks the
  publish (QoS 1 PUBACK), which is what ntfy's `delivered` also means: the
  delivery path took it, not that anyone saw it.
- `acknowledged` (epoch seconds) when the person dismissed the
  notification in the app. This is the far-end receipt the notifications
  settlement said only an app could give; whether an unacknowledged
  `alert` escalates is the automation's policy over this aspect, never
  this adapter's.
- `available` (bool) on both of a phone's entities, from the app's
  retained birth message and its last will. A phone out of coverage is
  unreachable, and an automation that escalates should be able to see it.
  Stale, not false: every other aspect keeps its last value.

Commands — the notifier's two commandable aspects, each a non-empty
string, published at QoS 1 and NOT retained (a retained alert would fire
again on every reconnect; the queueing is the app's persistent session,
which is the whole reason the broker carries this instead of ntfy). The
envelope's `actor` travels as `actor` so the phone can say who spoke.
There is no rate floor: unlike ntfy there is no third-party server to
protect, and the cooldown that is house policy lives in the automation.

Health events: drop/malformed-payload (undecodable or wrongly typed
input from the phone), drop/invalid-command (bad envelope, unknown
aspect, non-string or empty text), drop/unknown-device (a subtree no
entity file binds, first sight only). Discovery is the static
one-record-per-entity document with the aspect descriptor.
"""

import json
import os
import threading
import time

import homeostat
from homeostat import house, keys, mqtt

DEFAULT_BASE_TOPIC = "companion"

# Commandable aspect -> the leaf it publishes on. Both channels are one
# topic each: the severity split is the app's notification channel, as it
# is ntfy's priority.
CHANNELS = {"message": "message", "alert": "alert"}

# Optional position fields, homeostat names already.
POSITION_OPTIONAL = ("accuracy", "battery", "fixed_at")

PERSON_DESCRIPTOR = {
    "schema": 1,
    "groups": ["presence"],
    "fields": {
        "presence": {"label": "home", "kind": "boolean", "group": "presence"},
    },
}

NOTIFIER_DESCRIPTOR = {
    "schema": 1,
    "groups": ["delivery"],
    "fields": {
        "delivered": {"label": "last delivered (epoch s)", "kind": "number", "group": "delivery"},
        "acknowledged": {"label": "last seen (epoch s)", "kind": "number", "group": "delivery"},
    },
}


def _number(value):
    """The value if it is a JSON number, else None — bools are not."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    return value


def main():
    unit = os.environ[keys.ENV_UNIT]
    config = house.load_adapter(unit)
    endpoint = mqtt.parse_endpoint(config.endpoint)
    base = mqtt.base_topic(endpoint, DEFAULT_BASE_TOPIC)

    by_id = {e.id: e for e in config.entities}
    # A phone's entities, for the one topic that addresses the phone
    # itself rather than one of its faces.
    by_phone: dict[str, list] = {}
    for entity in config.entities:
        by_phone.setdefault(entity.id.split("/")[0], []).append(entity)

    session = homeostat.connect()

    seen_unknown: set[str] = set()
    available: dict[str, bool] = {}
    # PUBACK arrives on the network thread; the mid is the only handle
    # paho gives back, so the notifier waits for it there.
    pending: dict[int, object] = {}
    lock = threading.Lock()

    def drop(reason, **fields):
        session.health_event("drop", reason=reason, **fields)

    def put(entity, aspect, value):
        session.put_json(keys.state_key(entity.room, entity.name, aspect), value)

    def set_available(entity, value: bool) -> None:
        # On transition only (docs/adapters.md §7).
        if available.get(entity.name) == value:
            return
        available[entity.name] = value
        put(entity, "available", value)

    def on_phone_message(client, userdata, msg):
        rest = msg.topic[len(base) + 1 :].split("/")
        phone, leaf = rest[0], rest[-1]
        entity_id = "/".join(rest[:-1])
        try:
            payload = json.loads(msg.payload)
        except ValueError:
            drop("malformed-payload", topic=msg.topic)
            return

        if leaf == "available":
            entities = by_phone.get(phone)
            if not entities:
                if phone not in seen_unknown:
                    seen_unknown.add(phone)
                    drop("unknown-device", topic=msg.topic)
                return
            if not isinstance(payload, bool):
                drop("malformed-payload", topic=msg.topic)
                return
            for entity in entities:
                set_available(entity, payload)
            return

        entity = by_id.get(entity_id)
        if entity is None:
            # Once per subtree: a phone the owner has not bound yet keeps
            # publishing, and repeating the event per fix would bury the
            # health feed.
            if entity_id not in seen_unknown:
                seen_unknown.add(entity_id)
                drop("unknown-device", topic=msg.topic)
            return

        if leaf == "presence":
            if not isinstance(payload, bool):
                drop("malformed-payload", topic=msg.topic)
                return
            put(entity, "presence", payload)
        elif leaf == "position":
            lat = _number(payload.get("lat")) if isinstance(payload, dict) else None
            lon = _number(payload.get("lon")) if isinstance(payload, dict) else None
            if lat is None or lon is None:
                drop("malformed-payload", topic=msg.topic)
                return
            put(entity, "lat", lat)
            put(entity, "lon", lon)
            for field in POSITION_OPTIONAL:
                value = _number(payload.get(field))
                if value is not None:
                    put(entity, field, value)
        elif leaf == "ack":
            at = _number(payload)
            if at is None:
                drop("malformed-payload", topic=msg.topic)
                return
            put(entity, "acknowledged", at)

    def on_publish(client, userdata, mid, reason_code=None, properties=None):
        with lock:
            entity = pending.pop(mid, None)
        if entity is not None:
            put(entity, "delivered", time.time())

    topics = [
        (f"{base}/+/available", 1),
        (f"{base}/+/person/presence", 1),
        (f"{base}/+/person/position", 1),
        (f"{base}/+/notifier/ack", 1),
    ]
    client = mqtt.connect(endpoint, on_phone_message, topics)
    client.on_publish = on_publish

    def cmd_handler(entity):
        def handler(sample):
            key = str(sample.key_expr)
            aspect = key.split("/", 4)[4]
            try:
                payload = json.loads(sample.payload.to_bytes())
            except ValueError:
                drop("malformed-payload", key=key)
                return
            try:
                value = keys.parse_cmd_envelope(payload)
            except ValueError:
                drop("invalid-command", key=key)
                return
            if aspect not in CHANNELS or not isinstance(value, str) or not value.strip():
                drop("invalid-command", key=key, aspect=aspect, value=value)
                return
            actor = payload.get("actor")
            body = json.dumps(
                {
                    "text": value,
                    "actor": actor if isinstance(actor, str) and actor else unit,
                    "sent_at": time.time(),
                }
            )
            phone = entity.id.split("/")[0]
            with lock:
                info = client.publish(f"{base}/{phone}/notifier/{CHANNELS[aspect]}", body, qos=1)
                pending[info.mid] = entity

        return handler

    subscribers = [
        session.subscribe(expr, cmd_handler(e))
        for e in config.entities
        if e.capability == "notifier"
        for expr in keys.command_keyexprs(e)
    ]

    session.put_json(
        keys.discovery_key(unit),
        [
            {
                "id": e.id,
                "configured": True,
                "entity": e.name,
                "suggested": {
                    "capability": e.capability,
                    "features": ["alert", "acknowledged"] if e.capability == "notifier" else [],
                },
                "aspects": NOTIFIER_DESCRIPTOR if e.capability == "notifier" else PERSON_DESCRIPTOR,
            }
            for e in config.entities
        ],
    )

    session.ready()

    mqtt.wait_for_shutdown()

    # The MQTT loop stops first: an in-flight callback during teardown
    # would otherwise put on a closed zenoh session.
    client.loop_stop()
    client.disconnect()
    for sub in subscribers:
        sub.undeclare()
    session.close()


if __name__ == "__main__":
    main()

File: tests/companion.rs
//! Companion adapter integration tests: each scenario spawns a real
//! mosquitto broker on a free port plus the real supervisor on the
//! companion fixture house, and asserts on both buses — the phone's MQTT
//! subtree and the house bus.

mod common;

use std::time::Duration;

use serde_json::{json, Value};
use zenoh::sample::SampleKind;

use common::{
    assert_unit_contract, await_discovery, await_states, cache_read, expect_drop_event,
    expect_states, matched_publisher, next_event, Mosquitto, Mqtt, StateSub, Supervisor,
};

const FIXTURE: &str = "tests/fixture_house_companion";
const PORT_ENV: &str = "HOMEOSTAT_TEST_MQTT_PORT";
const EVENT_KEY: &str = "home/health/companion/event";
const ALICE_MESSAGE: &str = "home/cmd/person/alice_phone/message";
const ALICE_ALERT: &str = "home/cmd/person/alice_phone/alert";

/// Spawns broker + supervisor on the fixture and waits for the adapter's
/// liveliness token (generous timeout: first run resolves the uv env).
async fn setup() -> (Mosquitto, Supervisor, zenoh::Session) {
    let mosquitto = Mosquitto::spawn();
    let port = mosquitto.port.to_string();
    let sup = Supervisor::spawn_with_env(FIXTURE, &[(PORT_ENV, &port)]);
    let observer = sup.observer().await;
    let token_sub = observer
        .liveliness()
        .declare_subscriber("home/health/companion/alive")
        .history(true)
        .await
        .expect("liveliness subscriber");
    let token = tokio::time::timeout(Duration::from_secs(60), token_sub.recv_async())
        .await
        .expect("adapter liveliness token within 60s")
        .expect("liveliness stream open");
    assert_eq!(token.kind(), SampleKind::Put);
    (mosquitto, sup, observer)
}

async fn event_sub(observer: &zenoh::Session) -> StateSub {
    observer
        .declare_subscriber(EVENT_KEY)
        .await
        .expect("event subscriber")
}

fn wish(text: Value, actor: &str) -> Value {
    json!({"value": text, "priority": "automation", "actor": actor})
}

fn body(payload: &[u8]) -> Value {
    serde_json::from_slice(payload).expect("the phone's payload is JSON")
}

/// Polls the core mirror until `key` carries a number, and returns it —
/// the shape for a timestamp no test can predict.
async fn await_number(observer: &zenoh::Session, key: &str) -> f64 {
    let deadline = tokio::time::Instant::now() + Duration::from_secs(20);
    loop {
        if let Some(value) = cache_read(observer, key).await {
            if let Some(number) = value.as_f64() {
                return number;
            }
        }
        assert!(
            tokio::time::Instant::now() < deadline,
            "{key} never carried a number"
        );
        tokio::time::sleep(Duration::from_millis(200)).await;
    }
}

/// (a) The phone's two upward channels: a geofence transition becomes
/// `presence` on the person, and an opt-in position fix fans out to the
/// scalar aspects — the optional fields only when the fix carries them.
#[tokio::test(flavor = "multi_thread")]
async fn presence_and_position_translate_to_bus_state() {
    let (mosquitto, mut sup, observer) = setup().await;
    let state_sub = observer
        .declare_subscriber("home/state/**")
        .await
        .expect("state subscriber");
    let mut mqtt = Mqtt::connect(mosquitto.port, "test-state").await;

    mqtt.publish("companion/alice/person/presence", "true")
        .await;
    expect_states(
        &state_sub,
        &[("home/state/person/alice/presence", json!(true))],
    )
    .await;

    mqtt.publish(
        "companion/alice/person/position",
        r#"{"lat":59.33,"lon":18.06,"accuracy":12,"battery":87,"fixed_at":1752600000}"#,
    )
    .await;
    expect_states(
        &state_sub,
        &[
            ("home/state/person/alice/lat", json!(59.33)),
            ("home/state/person/alice/lon", json!(18.06)),
            ("home/state/person/alice/accuracy", json!(12)),
            ("home/state/person/alice/battery", json!(87)),
            ("home/state/person/alice/fixed_at", json!(1752600000)),
        ],
    )
    .await;

    // A fix without the optional fields publishes lat/lon and invents
    // nothing: the next presence sentinel arrives with no accuracy
    // between it and the fix.
    mqtt.publish(
        "companion/alice/person/position",
        r#"{"lat":1.0,"lon":2.0}"#,
    )
    .await;
    mqtt.publish("companion/alice/person/presence", "false")
        .await;
    expect_states(
        &state_sub,
        &[
            ("home/state/person/alice/lat", json!(1.0)),
            ("home/state/person/alice/presence", json!(false)),
        ],
    )
    .await;

    assert_unit_contract(&mut sup, &observer, "companion").await;
}

/// (b) A message and an alert reach the phone's own notifier topics,
/// carrying the envelope's actor, and the broker's QoS 1 ack publishes as
/// `delivered` — the delivery path took it, never a human's receipt.
#[tokio::test(flavor = "multi_thread")]
async fn notifications_reach_the_phone_and_publish_delivered() {
    let (mosquitto, mut sup, observer) = setup().await;
    let mut mqtt = Mqtt::connect(mosquitto.port, "test-notify").await;
    mqtt.subscribe("companion/alice/notifier/#").await;

    let message = matched_publisher(&observer, ALICE_MESSAGE).await;
    message
        .put(wish(json!("Irrigation skipped: it rained"), "irrigation").to_string())
        .await
        .expect("put");
    let (topic, payload) = mqtt
        .next_message(Duration::from_secs(20))
        .await
        .expect("message on the phone's topic");
    assert_eq!(topic, "companion/alice/notifier/message");
    let sent = body(&payload);
    assert_eq!(sent["text"], "Irrigation skipped: it rained");
    assert_eq!(sent["actor"], "irrigation");
    assert!(sent["sent_at"].is_number(), "{sent}");

    let alert = matched_publisher(&observer, ALICE_ALERT).await;
    alert
        .put(wish(json!("Motion in the hall and nobody home"), "intrusion").to_string())
        .await
        .expect("put");
    let (topic, payload) = mqtt
        .next_message(Duration::from_secs(20))
        .await
        .expect("alert on the phone's topic");
    assert_eq!(topic, "companion/alice/notifier/alert");
    assert_eq!(body(&payload)["text"], "Motion in the hall and nobody home");

    // delivered is the broker's ack, so it lands without the phone ever
    // saying anything back.
    let delivered = await_number(&observer, "home/state/person/alice_phone/delivered").await;
    assert!(delivered > 0.0, "delivered is an epoch: {delivered}");

    assert_unit_contract(&mut sup, &observer, "companion").await;
}

/// (c) The far-end receipt and the phone's own liveness: an ack becomes
/// `acknowledged` on the notifier, and the app's birth/last-will topic
/// flips `available` on BOTH of that phone's entities, on transition.
#[tokio::test(flavor = "multi_thread")]
async fn ack_and_availability_reach_both_faces() {
    let (mosquitto, mut sup, observer) = setup().await;
    let state_sub = observer
        .declare_subscriber("home/state/**")
        .await
        .expect("state subscriber");
    let mut mqtt = Mqtt::connect(mosquitto.port, "test-ack").await;

    mqtt.publish("companion/alice/notifier/ack", "1752600123")
        .await;
    expect_states(
        &state_sub,
        &[(
            "home/state/person/alice_phone/acknowledged",
            json!(1752600123),
        )],
    )
    .await;

    mqtt.publish("companion/alice/available", "true").await;
    await_states(
        &observer,
        &state_sub,
        &[
            ("home/state/person/alice/available", json!(true)),
            ("home/state/person/alice_phone/available", json!(true)),
        ],
    )
    .await;

    // What a last will looks like on the wire when the phone drops off.
    mqtt.publish("companion/alice/available", "false").await;
    await_states(
        &observer,
        &state_sub,
        &[
            ("home/state/person/alice/available", json!(false)),
            ("home/state/person/alice_phone/available", json!(false)),
        ],
    )
    .await;

    assert_unit_contract(&mut sup, &observer, "companion").await;
}

/// (d) Everything the phone can say wrong drops with a health event and
/// leaves the adapter translating.
#[tokio::test(flavor = "multi_thread")]
async fn bad_phone_input_drops_with_health_event() {
    let (mosquitto, mut sup, observer) = setup().await;
    let events = event_sub(&observer).await;
    let state_sub = observer
        .declare_subscriber("home/state/**")
        .await
        .expect("state subscriber");
    let mut mqtt = Mqtt::connect(mosquitto.port, "test-bad").await;

    mqtt.publish("companion/alice/person/presence", "certainly not json")
        .await;
    expect_drop_event(&events, "malformed-payload").await;

    // A presence that is not a boolean, a fix without lat, an ack that is
    // not a number, an availability that is not a boolean.
    mqtt.publish("companion/alice/person/presence", r#""home""#)
        .await;
    expect_drop_event(&events, "malformed-payload").await;
    mqtt.publish("companion/alice/person/position", r#"{"lon":18.06}"#)
        .await;
    expect_drop_event(&events, "malformed-payload").await;
    mqtt.publish("companion/alice/notifier/ack", r#""just now""#)
        .await;
    expect_drop_event(&events, "malformed-payload").await;
    mqtt.publish("companion/alice/available", r#""yes""#).await;
    expect_drop_event(&events, "malformed-payload").await;

    // Still translating.
    mqtt.publish("companion/alice/person/presence", "true")
        .await;
    expect_states(
        &state_sub,
        &[("home/state/person/alice/presence", json!(true))],
    )
    .await;

    sup.shutdown();
    drop(mosquitto);
}

/// (e) A subtree no entity file binds reports once, not per publish: an
/// unbound phone keeps publishing forever and discovery is not the health
/// feed's job.
#[tokio::test(flavor = "multi_thread")]
async fn an_unbound_phone_reports_once() {
    let (mosquitto, mut sup, observer) = setup().await;
    let events = event_sub(&observer).await;
    let mut mqtt = Mqtt::connect(mosquitto.port, "test-unbound").await;

    for _ in 0..3 {
        mqtt.publish("companion/zoe/person/presence", "true").await;
    }
    let event = next_event(&events).await;
    assert_eq!(event["reason"], json!("unknown-device"));
    assert_eq!(event["topic"], json!("companion/zoe/person/presence"));

    // A sentinel published last, asserted strictly: any further
    // unknown-device from zoe's remaining two publishes would precede it.
    mqtt.publish("companion/alice/person/presence", "not json")
        .await;
    assert_eq!(
        next_event(&events).await["reason"],
        json!("malformed-payload"),
        "an unbound phone must not report every publish"
    );

    sup.shutdown();
    drop(mosquitto);
}

/// (f) What never reaches the phone: a non-string value, an empty string,
/// an unknown aspect, an envelope-less payload.
#[tokio::test(flavor = "multi_thread")]
async fn invalid_commands_never_reach_the_phone() {
    let (mosquitto, mut sup, observer) = setup().await;
    let events = event_sub(&observer).await;
    let mut mqtt = Mqtt::connect(mosquitto.port, "test-invalid").await;
    mqtt.subscribe("companion/alice/notifier/#").await;

    let alert = matched_publisher(&observer, ALICE_ALERT).await;
    alert
        .put(wish(json!(42), "x").to_string())
        .await
        .expect("put");
    expect_drop_event(&events, "invalid-command").await;
    alert
        .put(wish(json!("   "), "x").to_string())
        .await
        .expect("put");
    expect_drop_event(&events, "invalid-command").await;
    alert
        .put(json!("bare string, no envelope").to_string())
        .await
        .expect("put");
    expect_drop_event(&events, "invalid-command").await;

    let odd = matched_publisher(&observer, "home/cmd/person/alice_phone/ring").await;
    odd.put(wish(json!("hello"), "x").to_string())
        .await
        .expect("put");
    let event = expect_drop_event(&events, "invalid-command").await;
    assert_eq!(event["aspect"], "ring");

    // The sentinel is the first thing the phone hears.
    let message = matched_publisher(&observer, ALICE_MESSAGE).await;
    message
        .put(wish(json!("the only one"), "x").to_string())
        .await
        .expect("put");
    let (topic, payload) = mqtt
        .next_message(Duration::from_secs(20))
        .await
        .expect("the sentinel");
    assert_eq!(topic, "companion/alice/notifier/message");
    assert_eq!(body(&payload)["text"], "the only one");

    sup.shutdown();
    drop(mosquitto);
}

/// (g) Discovery: one record per bound entity, each with the aspect
/// descriptor its face needs.
#[tokio::test(flavor = "multi_thread")]
async fn discovery_describes_both_faces() {
    let (_mosquitto, mut sup, observer) = setup().await;
    let doc = await_discovery(&observer, "companion", |d| {
        d.as_array().is_some_and(|records| records.len() == 2)
    })
    .await;
    let records = doc.as_array().expect("discovery is an array");

    let person = records
        .iter()
        .find(|r| r["id"] == json!("alice/person"))
        .expect("person record");
    assert_eq!(person["configured"], json!(true));
    assert_eq!(person["entity"], json!("alice"));
    assert_eq!(
        person["suggested"],
        json!({"capability": "person", "features": []})
    );
    assert_eq!(person["aspects"]["fields"]["presence"]["kind"], "boolean");

    let notifier = records
        .iter()
        .find(|r| r["id"] == json!("alice/notifier"))
        .expect("notifier record");
    assert_eq!(notifier["entity"], json!("alice_phone"));
    assert_eq!(
        notifier["suggested"],
        json!({"capability": "notifier", "features": ["alert", "acknowledged"]})
    );
    let fields = &notifier["aspects"]["fields"];
    assert_eq!(fields["delivered"]["kind"], "number");
    assert_eq!(fields["acknowledged"]["kind"], "number");

    sup.shutdown();
}
